swift/apple/FirezoneKit/Sources/FirezoneKit/Managers/TunnelManager.swift

```diff
@@ -0,0 +1,259 @@
+    guard let manager = manager,
+          let protocolConfiguration = manager.protocolConfiguration as? NETunnelProviderProtocol,
+          let providerConfiguration = protocolConfiguration.providerConfiguration as? [String: String]
+    else {
+      Log.app.error("Manager doesn't seem initialized. Can't save settings.")
+      throw TunnelManagerError.cannotSaveIfMissing
+    }
+
+    // Save and reload tunnel configuration
+    var newProviderConfiguration = settings.toProviderConfiguration()
+
+    // Don't clobber existing actorName
+    newProviderConfiguration[TunnelManagerKeys.actorName] = providerConfiguration[TunnelManagerKeys.actorName]
+    protocolConfiguration.providerConfiguration = newProviderConfiguration
+    protocolConfiguration.serverAddress = settings.apiURL
+    manager.protocolConfiguration = protocolConfiguration
+
+    // We always set this to true when starting the tunnel in case our tunnel
+    // was disabled by the system for some reason.
+    manager.isEnabled = true
+
+    try await manager.saveToPreferences()
+    try await manager.loadFromPreferences()
+  }
+
+  func start(token: String? = nil) {
+    var options: [String: NSObject]?
+
+    if let token = token {
+      options = ["token": token as NSObject]
+    }
+
+    do {
+      try session().startTunnel(options: options)
+    } catch {
+      Log.app.error("Error starting tunnel: \(error)")
+    }
+  }
+
+  func stop(clearToken: Bool = false) {
+    if clearToken {
+      do {
+        try session().sendProviderMessage("signOut".data(using: .utf8)!) { _ in
+          self.session().stopTunnel()
+        }
+      } catch {
+        Log.app.error("\(#function): \(error)")
+      }
+    } else {
+      session().stopTunnel()
+    }
+  }
+
+  func fetchResources(callback: @escaping (Data) -> Void) {
+    guard session().status == .connected else { return }
+
+    do {
+      try session().sendProviderMessage(resourceListHash) { data in
+        if let data = data {
+          self.resourceListHash = Data(SHA256.hash(data: data))
+          self.resourcesListCache = data
+        }
+
+        callback(self.resourcesListCache)
+      }
+    } catch {
+      Log.app.error("Error: sendProviderMessage: \(error)")
+    }
+  }
+
+  private func session() -> NETunnelProviderSession {
+    guard let manager = manager,
+          let session = manager.connection as? NETunnelProviderSession
+    else { fatalError("Could not cast tunnel connection to NETunnelProviderSession!") }
+
+    return session
+  }
+
+  // Subscribe to system notifications about our VPN status changing
+  // and let our handler know about them.
+  private func setupTunnelObservers() {
+    Log.app.log("\(#function)")
+
+    for task in tunnelObservingTasks {
+      task.cancel()
+    }
+    tunnelObservingTasks.removeAll()
+
+    tunnelObservingTasks.append(
+      Task {
+        for await notification in NotificationCenter.default.notifications(
```

Comment:
I removed the handler introduced in the last big refactor that handles configuration change -- I tried hitting it manually and it wouldn't fire, seems to be a cobweb in Apple's API